### PR TITLE
PLAT-226 test api SM

### PR DESCRIPTION
**Pull request by ruz**

### CODE REVIEW

File: ledger-core/v2/logicrunner/sm_request/call_result.go [hunk 1]
```diff
@@ -36,7 +38,18 @@ func (s *StateMachineCallResult) Init(ctx smachine.InitializationContext) smachi
 }
 
 func (s *StateMachineCallResult) stepProcess(ctx smachine.ExecutionContext) smachine.StateUpdate {
-	link, bgin := ctx.GetPublishedGlobalAliasAndBargeIn("TODO: some key")
+	var barginAlias string
+	switch s.Payload.CallType {
+	case payload.CTMethod:
+		barginAlias = "waiting for call result" // TODO make better
+	case payload.CTConstructor:
+		outgoingRef := *insolar.NewGlobalReference(*s.Payload.Caller.GetLocal(), s.Payload.CallOutgoing)
```
[the remark this answers is not in the document]
Comment (aalekseev-ru) [reply ledger-core/v2/logicrunner/sm_request/call_result.go:46]: func NewGlobalReference(local ID, base ID) *Reference

File: ledger-core/v2/logicrunner/sm_request/call_result.go [hunk 2]
```diff
@@ -36,7 +38,18 @@ func (s *StateMachineCallResult) Init(ctx smachine.InitializationContext) smachi
 }
 
 func (s *StateMachineCallResult) stepProcess(ctx smachine.ExecutionContext) smachine.StateUpdate {
-	link, bgin := ctx.GetPublishedGlobalAliasAndBargeIn("TODO: some key")
+	var barginAlias string
```
[the remark this answers is not in the document]
Comment (aalekseev-ru) [reply ledger-core/v2/logicrunner/sm_request/call_result.go:41]: fiixed

File: ledger-core/v2/logicrunner/sm_test_api/sm.go [hunk 3]
```diff
@@ -27,6 +28,11 @@ type TestApiCallSMDeclaration struct {
 	smachine.StateMachineDeclTemplate
 }
 
+func GetApiCaller() *insolar.Reference {
+	caller, _ := insolar.NewObjectReferenceFromString("insolar:1F5rFXwPMobzCzgQ7DeXwLOJj7VPaKmQMuc2ktAAAAAQ") // choosed by fairly random
```
[the remark this answers is not in the document]
Comment (aalekseev-ru) [reply ledger-core/v2/logicrunner/sm_test_api/sm.go:32]: I used it the same way as it used in proxy. It's easier for reading and maintaining.

[the remark this answers is not in the document]
Comment (aalekseev-ru) [reply ledger-core/v2/logicrunner/sm_test_api/sm.go:32]: I didn't get a difference for the first time, fixed.

File: ledger-core/v2/go.mod [hunk 4]
```diff
@@ -19,7 +19,7 @@ require (
 	github.com/golang/protobuf v1.3.2
 	github.com/google/gofuzz v1.0.0
 	github.com/google/gops v0.3.6
-	github.com/google/uuid v1.1.1
+	github.com/google/uuid v1.1.1 // indirect
```
Comment (sexyhamster) [inline ledger-core/v2/go.mod:22]: run go mod tidy. I'm not sure why this dependency keep toggling, but it should be direct

File: ledger-core/v2/go.mod [hunk 5]
```diff
@@ -19,7 +19,6 @@ require (
 	github.com/golang/protobuf v1.3.2
 	github.com/google/gofuzz v1.0.0
 	github.com/google/gops v0.3.6
-	github.com/google/uuid v1.1.1
```
Comment (sexyhamster) [inline ledger-core/v2/go.mod:22]: still very strange, we use uuid in various ways

Comment (aalekseev-ru) [reply ledger-core/v2/go.mod:22]: it's result of tidy

Comment (aalekseev-ru) [reply ledger-core/v2/go.mod]: @sexyhamster fixed with running go.mod tidy on fresh master

**Comment by ita-sammann** [comment]
When you will be merging this PR squash it plz

File: ledger-core/v2/logicrunner/common/conveyor_dispatcher.go [hunk 6]
```diff
@@ -108,3 +108,8 @@ func (c *conveyorDispatcher) Process(msg *message.Message) error {
 func NewConveyorDispatcher(conveyor *conveyor.PulseConveyor) dispatcher.Dispatcher {
 	return &conveyorDispatcher{conveyor: conveyor}
 }
+
+type TestApiCall struct {
```
Comment (bigbes) [inline ledger-core/v2/logicrunner/common/conveyor_dispatcher.go:112]: imho TestApiCall -> TestAPICall, if golinter is ok
TestAPICall should be moved from here, but i don't know where it should be.

File: ledger-core/v2/logicrunner/sm_test_api/sm.go [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+	// injected arguments
+	PulseSlot *conveyor.PulseSlot
+	sender    *s_sender.SenderServiceAdapter
+}
+
+type TestApiCallSMDeclaration struct {
+	smachine.StateMachineDeclTemplate
+}
+
+var ApiCaller, _ = insolar.NewObjectReferenceFromString("insolar:0AAABAnRB0CKuqXTeTfQNTolmyixqQGMJz5sVvW81Dng")
```
Comment (bigbes) [inline ledger-core/v2/logicrunner/sm_test_api/sm.go:35]: APICaller won't be NodeID? Or we have no NodeID right now?

File: ledger-core/v2/logicrunner/sm_request/call_result.go [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+	if !bgin.IsValid() {
+		// TODO: log
+		return ctx.Stop()
+	}
+
+	done := bgin.CallWithParam(s.Payload)
+	if !done {
+		// TODO: log
+		return ctx.Stop()
+	}
```
Comment (bigbes) [inline ledger-core/v2/logicrunner/sm_request/call_result.go:64]: We should differentiate impossible situations and no one is waiting for response.

- "link is empty" == no one is waiting
- "bargein is empty" == impossible situation (for now)
- "bargein is invalid" == no one is waiting anymore (we can remove that check, it is done automatically in CallWithParam)
- "bargein failed to call" == no one is waiting anymore

And we should behave differently in both that cases.

Comment (ruz) [reply ledger-core/v2/logicrunner/sm_request/call_result.go:64]: We talked about this with @cyraxred in the slack and agreed that most of these should result in error terminating of the state machine, not ctx.Stop()

File: ledger-core/v2/logicrunner/sm_test_api/sm.go (hunk 7 above)
Comment (aalekseev-ru) [reply ledger-core/v2/logicrunner/sm_test_api/sm.go:35]: I think right now it doesn't matter, can we win something to change it to NodeID?

File: ledger-core/v2/logicrunner/sm_request/call_result.go (hunk 8 above)
Comment (bigbes) [reply ledger-core/v2/logicrunner/sm_request/call_result.go]: @ruz "no one is waiting" - not an error, is it?

Comment (bigbes) [reply ledger-core/v2/logicrunner/sm_request/call_result.go]: ok, see it now. "Weird flex, but ok"

File: ledger-core/v2/logicrunner/sm_request/call_result.go [hunk 9]
```diff
@@ -45,22 +45,15 @@ func (s *StateMachineCallResult) stepProcess(ctx smachine.ExecutionContext) smac
...
+		return ctx.Error(throw.E("no one is waiting", struct{ smachine.SlotLink }{link}))
 	}
 	if bgin == nil {
-		// TODO: log
-		return ctx.Stop()
-	}
-	if !bgin.IsValid() {
-		// TODO: log
-		return ctx.Stop()
+		return ctx.Error(throw.E("impossible situation ", struct{ smachine.BargeInHolder }{bgin}))
```
Comment (bigbes) [inline ledger-core/v2/logicrunner/sm_request/call_result.go:51]: Replace with simple `ctx.Error(throw.Impossible())`, bgin is nil, so we don't need no info here

File: ledger-core/v2/logicrunner/sm_request/call_result.go [hunk 10]
```diff
@@ -45,22 +45,15 @@ func (s *StateMachineCallResult) stepProcess(ctx smachine.ExecutionContext) smac
 	outgoingRef := reference.NewGlobal(*s.Payload.Caller.GetLocal(), s.Payload.CallOutgoing)
 	link, bgin := ctx.GetPublishedGlobalAliasAndBargeIn(outgoingRef)
 	if link.IsZero() {
-		// TODO: log
-		return ctx.Stop()
+		return ctx.Error(throw.E("no one is waiting", struct{ smachine.SlotLink }{link}))
```
Comment (bigbes) [inline ledger-core/v2/logicrunner/sm_request/call_result.go:48]: No need to push link here, we should see it in log anyway

**Pull request closed without merging**